pkg/ucp/store/inmemory/client.go

```diff
@@ -0,0 +1,262 @@
+
+	return nil
+}
+
+// Query implements store.StorageClient.
+func (c *Client) Query(ctx context.Context, query store.Query, options ...store.QueryOptions) (*store.ObjectQueryResult, error) {
+	if ctx == nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+
+	err := query.Validate()
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: fmt.Sprintf("invalid argument. Query is invalid: %s", err.Error())}
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	result := &store.ObjectQueryResult{}
+	for _, entry := range c.resources {
+		// Check root scope.
+		if query.ScopeRecursive && !strings.HasPrefix(entry.rootScope, storeutil.NormalizePart(query.RootScope)) {
+			continue
+		} else if !query.ScopeRecursive && entry.rootScope != storeutil.NormalizePart(query.RootScope) {
+			continue
+		}
+
+		// Check resource type.
+		resourceType, err := storeutil.NormalizeResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
+		if entry.resourceType != storeutil.NormalizePart(resourceType) {
+			continue
+		}
+
+		// Check routing scope prefix (optional).
+		if query.RoutingScopePrefix != "" && !strings.HasPrefix(entry.routingScope, storeutil.NormalizePart(query.RoutingScopePrefix)) {
+			continue
+		}
+
+		// Check filters (optional).
+		match, err := entry.obj.MatchesFilters(query.Filters)
+		if err != nil {
+			return nil, err
+		}
+		if !match {
+			continue
+		}
+
+		result.Items = append(result.Items, entry.obj)
+	}
+
+	return result, nil
+}
+
+// Save implements store.StorageClient.
+func (c *Client) Save(ctx context.Context, obj *store.Object, options ...store.SaveOptions) error {
+	if ctx == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+	if obj == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj' is required"}
+	}
+
+	parsed, err := resources.Parse(obj.ID)
+	if err != nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'obj.ID' must be a valid resource id"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	config := store.NewSaveConfig(options...)
+
+	entry, ok := c.resources[strings.ToLower(normalized.String())]
+	if !ok && config.ETag != "" {
+		return &store.ErrConcurrency{}
+	} else if ok && config.ETag != "" && config.ETag != entry.obj.ETag {
+		return &store.ErrConcurrency{}
+	} else if !ok {
+		// New entry, initialize it.
+		entry.rootScope = storeutil.NormalizePart(normalized.RootScope())
+		entry.resourceType = storeutil.NormalizePart(normalized.Type())
+		entry.routingScope = storeutil.NormalizePart(normalized.RoutingScope())
+	}
+
+	raw, err := json.Marshal(obj.Data)
+	if err != nil {
+		return err
+	}
+
+	obj.ETag = etag.New(raw)
+	entry.obj = *obj
+	c.resources[strings.ToLower(normalized.String())] = entry
+
+	return nil
+}
+
+// Clear can be used to clear all stored data.
+func (c *Client) Clear() {
```

Comment:
This can be useful for 409 errors :)

Reply:
I don't get it, but if you're psyched them so am I 😆